osgp/shared/shared/src/main/java/org/opensmartgridplatform/shared/security/providers/HsmEncryptionProvider.java

```diff
@@ -45,63 +45,77 @@
             throw new EncrypterException("Could not read keystore", e);
         }
     }
 
-    public Secret decrypt(EncryptedSecret secret, String keyReference) {
+    @Override
+    public byte[] decrypt(final EncryptedSecret secret, final String keyReference) throws EncrypterException {
 
-        Secret decryptedSecret = super.decrypt(secret, keyReference);
+        byte[] decryptedSecret = super.decrypt(secret, keyReference);
 
-        byte[] decryptedSecretBytes = decryptedSecret.getSecret();
+        if (decryptedSecret.length > KEY_LENGTH) {
 
-        if (decryptedSecretBytes.length > KEY_LENGTH) {
+            final byte[] truncatedDecryptedSecretBytes = Arrays.copyOfRange(decryptedSecret, 0,
+                    decryptedSecret.length-16);
 
-            byte[] truncatedDecryptedSecretBytes = Arrays.copyOfRange(decryptedSecretBytes, 0,
-                    decryptedSecretBytes.length-16);
-
-            LOGGER.trace("Truncating decrypted key from " + Hex.encodeHexString(decryptedSecretBytes) + " to " +
+            LOGGER.trace("Truncating decrypted key from " + Hex.encodeHexString(decryptedSecret) + " to " +
                             Hex.encodeHexString(truncatedDecryptedSecretBytes));
 
-            decryptedSecret = new Secret(truncatedDecryptedSecretBytes);
+            return truncatedDecryptedSecretBytes;
         }
 
         return decryptedSecret;
     }
 
-    protected Cipher getCipher() {
+    @Override
+    public byte[] generateAes128BitsSecret(String keyReference) throws EncrypterException {
+        //byte[] newSecret = new byte[128/8];
+        try {
+            //SecureRandom.getInstance("SHA1PRNG",PROVIDER).nextBytes(newSecret);
+            return this.encrypt(KeyGenerator.getInstance("AES").generateKey().getEncoded(),keyReference).getSecret();
+        } catch (NoSuchAlgorithmException exc) {
+            throw new EncrypterException("Could not generate secret", exc);
+        }
+        //return newSecret;
+    }
+
+    @Override
+    protected Cipher getCipher() throws EncrypterException {
         try {
             return Cipher.getInstance(ALGORITHM, PROVIDER);
-        } catch (NoSuchPaddingException | NoSuchAlgorithmException | NoSuchProviderException e) {
+        } catch (final NoSuchPaddingException | NoSuchAlgorithmException | NoSuchProviderException e) {
             throw new EncrypterException("Could not get cipher", e);
         }
     }
 
     /**
-     * This method reads the 'actual' encryption key (from the database).
-     * Normally this is the key start isValidFrom(now) and isValidUntil(now).
+     * This method reads the encryption key specified by keyReference from the Hsm.
      *
      * @return the key that must be used for encryption/decryption
      */
-    protected Key getSecretEncryptionKey(String keyReference, int cipherMode) {
+    @Override
+    protected Key getSecretEncryptionKey(final String keyReference, final int cipherMode) throws EncrypterException {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of thrown exception 'org.opensmartgridplatform.shared.exceptionhandling.EncrypterException' which is a runtime exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)